---
title: TAP-seq permits efficient identification of cell types and differentiation
  states at very low read depths (Figure 4)
author: "Schraivogel, Gschwind et al."
date: "02/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
options(bitmapType = "cairo")
```

This vignette reproduces the analysis in Figure 4 of Schraivogel, Gschwind, et al. 

## Data

As whole transcriptome reference data for bone marrow cell types, we use the data by Baccin et al., Nature Cell Biology in press.

```{sh getReferenceData, eval = F}
mkdir data
wget https://nicheview.shiny.embl.de/RNAMagnetDataBundle.zip
unzip -j -d data RNAMagnetDataBundle.zip

rm RNAMagnetDataBundle.zip
```

We also download the count matrices for the targeted seq expeirments of total and c-Kit+ bone marrow. Creation of count matrices from raw files provided on GEO can be reproduced using XXX.

```{sh getCountData, eval =F}
wget -O data/dge_TAPtotalBM.txt.gz http://steinmetzlab.embl.de/TAPdata/dge_TAPtotalBM.txt.gz
wget -O data/dge_TAPkitBM.txt.gz http://steinmetzlab.embl.de/TAPdata/dge_TAPkitBM.txt.gz
```

Data were downsampeld at the level of raw reads to realistically simulate a lower sequencing depth. The downsampling can be reproduced from the files provided on GEO using XXX; alternatively we provide the results:

```{sh getDownsampledData, eval = F}
wget http://steinmetzlab.embl.de/TAPdata/TAPkitBM.zip
wget http://steinmetzlab.embl.de/TAPdata/TAPtotalBM.zip
wget http://steinmetzlab.embl.de/TAPdata/WholeKitBM.zip
wget http://steinmetzlab.embl.de/TAPdata/WholeTotalBM.zip

unzip -j -d data/TAPkitBM TAPkitBM.zip
unzip -j -d data/TAPtotalBM TAPtotalBM.zip
unzip -j -d data/WholeKitBM WholeKitBM.zip
unzip -j -d data/WholeTotalBM WholeTotalBM.zip

rm TAPkitBM.zip TAPtotalBM.zip WholeKitBM.zip WholeTotalBM.zip

```

## Compute environment

The following packages are required:

```{r pressure}
.libPaths("/g/steinmetz/velten/Software/RLibs-seurat3/")
require(Seurat)
require(ggplot2)
require(parallel)
require(plyr)
require(irr)
require(pheatmap)
require(mclust)
```

## Setup of basic Seurat objects

For TAP-seq, raw count matrixes are loaded into R and processed into a Seurat object using default settings.

```{r setupTAPseurat, results = "hide"}
dir <- "data"
files <- c("dge_TAPtotalBM.txt.gz")

DGE <-
  lapply(file.path(dir, files),
    function(x) read.table(gzfile(x),header = T,row.names = 1)
  )
names(DGE) <- gsub("dge_(.+)\\.txt.gz","\\1",files)
DGE <-
  lapply(names(DGE), function(n) {
    x <- DGE[[n]]
    colnames(x) <- paste(n, colnames(x), sep = "-")
    x
  })

common.genes <- rownames(DGE[[1]])
if (length(DGE) > 1)
  for (i in 2:length(DGE))
    common.genes <- intersect(common.genes, rownames(DGE[[i]]))

DGE <- lapply(DGE, function(x)
  x[common.genes,])
DGE <- do.call(cbind, DGE)


#try simple Seurat workflow 
TAPseq <- CreateSeuratObject(DGE, min.features = 10)
TAPseq <- NormalizeData(object = TAPseq)
TAPseq <- FindVariableFeatures(object = TAPseq)
TAPseq <- ScaleData(object = TAPseq)
TAPseq <- RunPCA(object = TAPseq)
TAPseq <- RunTSNE(object = TAPseq)
TAPseq <- FindNeighbors(object = TAPseq)
TAPseq <- FindClusters(object = TAPseq)
TAPseq$cluster <- Idents(TAPseq)
```

Reference data is simply loaded into R as a Seurat object. Only cell types abundantly present in c-Kit+ and total bone marrow are used for the reference.

```{r setupRefseurat}
load("data/NicheData10x.rda")
```

```{r collorscheme, echo=F}

NicheDataColors <-
  c(Erythroblasts = "#bc7c7c", Chondrocytes = "#a6c7f7", Osteoblasts = "#0061ff", 
    `Fibro/Chondro p.` = "#70a5f9", `pro-B` = "#7b9696", `Arteriolar ECs` = "#b5a800", 
    `B cell` = "#000000", `large pre-B.` = "#495959", `Sinusoidal ECs` = "#ffee00", 
    Fibroblasts = "#70a5f9", `Endosteal fibro.` = "#264570", `Arteriolar fibro.` = "#567fba", 
    `Stromal fibro.` = "#465f82", `small pre-B.` = "#323d3d", `Adipo-CAR` = "#ffb556", 
    `Ng2+ MSCs` = "#ab51ff", Neutrophils = "#1f7700", `T cells` = "#915400", 
    `NK cells` = "#846232", `Schwann cells` = "#ff00fa", `Osteo-CAR` = "#ff0000", 
    `Dendritic cells` = "#44593c", Myofibroblasts = "#dddddd", Monocytes = "#8fff68", 
    `Smooth muscle` = "#ff2068", `Ery prog.` = "#f9a7a7", `Mk prog.` = "#f9e0a7", 
    `Ery/Mk prog.` = "#f9cda7", `Gran/Mono prog.` = "#e0f9a7", `Neutro prog.` = "#c6f9a7", 
    `Mono prog.` = "#f4f9a7", LMPPs = "#a7f9e9", `Eo/Baso prog.` = "#a7b7f9", 
    HSPCs = "#c6f9a7")


```

## Unsupervised analysis of downsampled data

The following function reads in the downsampled data and performs Seurat clustering as for the non-downsampled data set. Rand indeces are then computed to compare the different cluster partitions. Downsampling is performed such that the average reads per cell is sampled to a given amount.

```{r runUnsupervisedDownsampling, fig.width=10,fig.height=10, results = "hide"}


UseCells <- subset(NicheData10x,metadata....experiment.. %in% c("2017_9_totalBM") )
UseCells$predicted.id <- Idents(UseCells)
UseCells <- FindVariableFeatures(UseCells)
UseCells <- RunPCA(UseCells)
UseCells <- FindNeighbors(UseCells)
UseCells <- FindClusters(UseCells)
newnames <- gsub("2018_2_HSPC_","WholeKitBM-",colnames(UseCells))
newnames <- gsub("2017_9_totalBM_","WholeTotalBM-",newnames)
names(newnames) <- colnames(UseCells)
UseCells <- RenameCells(UseCells, new.names = newnames)

source("functions/function_cellTypeDownsampling.unsupervised.R")

tap.average <-   getDS.wrapper("TAP","perSample","average", reference = TAPseq)

whole.average <- getDS.wrapper("Whole","perSample","average")

average <- rbind(tap.average, whole.average) #, whole.target.average
  
```


```{r figure 4d, fig.width=5, fig.height=4}

#average <- subset(average, !(reads > 2000 & experiment =="TAP"))
qplot(x = reads, y= kappa, color = experiment, data=subset(average, panel =="perSample"), log="x") + xlab("Average reads per cell") + ylab("Adjusted Rand Index") + theme_bw() + theme(panel.grid = element_blank(),legend.position = c(0.79,0.3), legend.background = element_rect(fill = NA)) + scale_color_discrete(name = "Method",labels = c("TAP"="TAP-seq", "Whole" = "10x")) + geom_smooth(se=F, size =0.5,linetype =2, span=2)

```


```{r figure 4e, fig.width=4, fig.height=4}

fit.tap <- loess(kappa ~ log10(reads), data= subset(average, panel =="perSample" & experiment == "TAP"), span=2)
fit.whole <- loess(kappa ~ log10(reads), data= subset(average, panel =="perSample"  & experiment == "Whole"), span=2)
lookup <- data.frame(reads = 10^seq(2,5.5,by=0.001),
                     kappa.tap = predict(fit.tap, newdata = data.frame(reads = 10^seq(2,5.5,by=0.001))),
                     kappa.whole = predict(fit.whole, newdata = data.frame(reads = 10^seq(2,5.5,by=0.001))))
cost <-data.frame(kappa = seq(0.6,0.8,by =0.001),
                  reads.tap = sapply(seq(0.6,0.8,by =0.001), function(x) lookup$reads[which.min(abs(lookup$kappa.tap - x))]),
                  reads.whole = sapply(seq(0.6,0.8,by =0.001), function(x) lookup$reads[which.min(abs(lookup$kappa.whole - x))]))
cost$savings <- cost$reads.whole / cost$reads.tap

qplot(x = kappa, y= savings, data=cost, geom="line") + xlab("Adjusted Rand Index") + ylab("Fold cost reduction") + theme_bw() + theme(panel.grid = element_blank()) + scale_y_continuous(limits = c(0,15))
```

## Supervised analysis of downsampled data

In ther context of supervised analyses, we first need a clean annotation reference, which we obtain by subsetting the Baccin et al dataset to only contain celltypes with abundant presence in total and c-Kit+ bone marrow:

```{r figure4b, fig.width=10,fig.height=8}
UseCells <- subset(NicheData10x, idents = c("Monocytes","Ery prog.","pro-B","T cells","Neutrophils","small pre-B.","Dendritic cells","NK cells","B cell"))

anchors <- FindTransferAnchors(reference = UseCells, query = TAPseq, dims = 1:15)
predictions <- TransferData(anchorset = anchors, refdata = Idents(UseCells), dims = 1:15)
TAPseq <- AddMetaData(object = TAPseq, metadata = predictions)
Idents(TAPseq) <- TAPseq$cluster

# 
mean_by_cluster <- lapply(as.character(unique(Idents(TAPseq))), function(x) {
  out <- data.frame( ge = apply(GetAssayData(TAPseq, slot = "data")[,Idents(TAPseq) == x],1,mean))
  out
})
mean_by_cluster <- do.call(cbind, mean_by_cluster)
colnames(mean_by_cluster) <- as.character(unique(Idents(TAPseq)))
metaclustering <- hclust(as.dist(1-cor(mean_by_cluster))) #create a pretty heatmap... for the identification of metaclusters


coldata <- data.frame(row.names = colnames(TAPseq),
                      id = factor(Idents(TAPseq), levels = metaclustering$labels[metaclustering$order]),
                      projected = TAPseq$predicted.id,
                      experiment = gsub("-.+","",colnames(TAPseq)))
ann_colors = list(projected = NicheDataColors[unique(as.character(coldata$projected))])
coldata <- coldata[order(coldata$id),]
require(pheatmap)
toplot <- GetAssayData(TAPseq, slot = "scale.data")[,rownames(coldata)]
toplot[toplot > 6] <- 6

pheatmap(toplot, cluster_cols = F, annotation_col = coldata,show_colnames = F, show_rownames = F, annotation_colors = ann_colors )
```

We then rename the clusters identified through unsupervised analyses of the TAP-seq dataset to match the names of the whole transcriptome dataset. Since T and NK cells are insufficiently sanpled in both datasets, these two cell types are merged.

Finally, again construct Seurat objects from the downsampled datasets, but this time, project on the non-downsampled reference.

```{r runSupervisedDownsampling, fig.width=4,fig.height=4, results = "hide"}

source("functions/function_cellTypeDownsampling.supervised.R")

#create an annotated version of the TAPseq dataset
TAPseq$predicted.id <- factor(TAPseq$predicted.id)
ident.table <- sapply(as.character(unique(Idents(TAPseq))), function(id) table(TAPseq$predicted.id[Idents(TAPseq)==id]))
renamer <- rownames(ident.table)[apply(ident.table,2,which.max)]
names(renamer) <- colnames(ident.table)
#renamer["11"] <- "large pre-B."
#renamer["15"] <- "Ery prog."
TAPseq <- RenameIdents(TAPseq, renamer)
TAPseq <- RenameIdents(TAPseq, "NK cells" = "T cells")

UseCells <- RenameIdents(object = UseCells, "NK cells" = "T cells")


mean_by_cluster <- lapply(as.character(unique(Idents(UseCells))), function(x) {
  out <- data.frame( ge = apply(GetAssayData(UseCells, slot = "data")[,Idents(UseCells) == x],1,mean))
  out
})
mean_by_cluster <- do.call(cbind, mean_by_cluster)
colnames(mean_by_cluster) <- as.character(unique(Idents(UseCells)))
metaclustering <- hclust(as.dist(1-cor(mean_by_cluster))) #create a pretty heatmap... for the identification of metaclusters


tap.average <-   getDS.wrapper("TAP","perSample","average", reference = TAPseq)
whole.average <- getDS.wrapper("Whole","perSample","average")

average <- rbind(tap.average, whole.average) #, whole.target.average

```

Also project deeply sequenced data on itsself. Find out how many reads each cell from the original datasets has.

```{r selfProject}

#project UseCells on itsself
self.anchors <- FindTransferAnchors(reference = UseCells, query = UseCells, dims = 1:15)
self.predictions <- TransferData(anchorset = self.anchors, refdata = Idents(UseCells), dims = 1:15)


self.k <- kappa2(
  data.frame(
    self.predictions$predicted.id,
    Idents(UseCells)
  )
)$value

reads.whole <- read.table(url("http://steinmetzlab.embl.de/TAPdata/WholeTotalBM.reads_per_cell_barcode.txt"))
rownames(reads.whole) <- paste0("2017_9_totalBM_", reads.whole$V2)

reads.whole <- rbind(reads.whole)


average <- rbind(average,
                 data.frame(experiment = "Whole",panel = "perSample", sampling = "average", kappa = self.k, reads = mean(reads.whole$V1[rownames(reads.whole) %in% colnames(UseCells)], na.omit = T), ncells = ncol(UseCells)))


#project TAPseq on itsself
self.anchors <- FindTransferAnchors(reference = TAPseq, query = TAPseq, dims = 1:15)
self.predictions <- TransferData(anchorset = self.anchors, refdata = Idents(TAPseq), dims = 1:15)


self.k <- kappa2(
  data.frame(
    self.predictions$predicted.id,
    Idents(TAPseq)
  )
)$value

reads.whole <- read.table(url("http://steinmetzlab.embl.de/TAPdata/TAPtotalBM.reads_per_cell_barcode.txt"))
rownames(reads.whole) <- paste0("TAPtotalBM-", reads.whole$V2)
reads.whole <- rbind(reads.whole)


average <- rbind(average,
                 data.frame(experiment = "TAP",panel = "perSample", sampling = "average", kappa = self.k, reads = mean(reads.whole$V1[rownames(reads.whole) %in% colnames(TAPseq)], na.omit = T), ncells = ncol(UseCells)))

```

Plot figure 4g

```{r figure4g, fig.width=4, fig.height=4}
qplot(x = reads, y=100* kappa, color = experiment, data=subset(average, panel =="perSample" & reads != 5500), log="x") + xlab("Average reads per cell") + ylab("% correctly classifed cells") + theme_bw() + theme(panel.grid = element_blank(),legend.position = c(0.7,0.3),axis.title.y = element_text(size=10)) + scale_color_discrete(name = "Method",labels = c("TAP"="TAP-seq", "Whole" = "10x")) + geom_smooth(se=F, size =0.5,linetype =2)

```

Plot figure 4h 
```{r figure4h, fig.width=4, fig.height=4}

fit.tap <- loess(kappa ~ log10(reads), data= subset(average, panel =="perSample" & reads != 5500 & experiment == "TAP"))
fit.whole <- loess(kappa ~ log10(reads), data= subset(average, panel =="perSample" & reads != 5500 & experiment == "Whole"))
lookup <- data.frame(reads = 10^seq(2,5.5,by=0.001),
                     kappa.tap = predict(fit.tap, newdata = data.frame(reads = 10^seq(2,5.5,by=0.001))),
                     kappa.whole = predict(fit.whole, newdata = data.frame(reads = 10^seq(2,5.5,by=0.001))))
cost <-data.frame(kappa = seq(0.75,0.93,by =0.001),
                  reads.tap = sapply(seq(0.75,0.93,by =0.001), function(x) lookup$reads[which.min(abs(lookup$kappa.tap - x))]),
                  reads.whole = sapply(seq(0.75,0.93,by =0.001), function(x) lookup$reads[which.min(abs(lookup$kappa.whole - x))]))
cost$savings <- cost$reads.whole / cost$reads.tap

qplot(x = 100*kappa, y= savings, data=cost, geom="line") + xlab("% correctly classified cells") + ylab("Fold cost reduction") + theme_bw() + theme(panel.grid = element_blank()) + scale_y_continuous(limits = c(0,15)) + scale_x_continuous(limits = c(75,90))
```

## Downsampled t-SNE

```{r tsne, width=4,height=4}
folders <- c( "WholeTotalBM")
folders <- file.path(dir, folders)
Whole100 <- getDS(folders, "perSample","average",100, output = "seurat",minFeature = 10)
Whole100 <- RunTSNE(Whole100, dims = 1:15, check_duplicates = FALSE)
plf <- data.frame(x = Embeddings(Whole100,"tsne")[,1],
                  y = Embeddings(Whole100,"tsne")[,2])
include <- colnames(UseCells)
include <- gsub("2018_2_HSPC_","WholeKitBM-",include)
include <- gsub("2017_9_totalBM_","WholeTotalBM-",include)
id <- Idents(UseCells); names(id) <- include
plf <- plf[intersect(include,rownames(plf)),]
plf$id <- id[rownames(plf)]
qplot(x=x,y=y,color=id,data=plf, size=I(0.5)) + scale_color_manual(values=NicheDataColors, guide=F) + theme_bw() + theme(panel.grid = element_blank(), axis.title = element_blank(),axis.text = element_blank(),axis.ticks = element_blank())


folders <- c("TAPtotalBM")
folders <- file.path(dir, folders)
TAP100 <- getDS(folders, "perSample","average",100, output = "seurat", ncells = 2108)
TAP100 <- RunTSNE(TAP100, dims = 1:15)
plf2 <- data.frame(x = Embeddings(TAP100,"tsne")[,1],
                   y = Embeddings(TAP100,"tsne")[,2],
                   id = TAP100$predicted.id)

qplot(x=x,y=y,color=id,data=plf2, size=I(0.5)) + scale_color_manual(values=NicheDataColors, guide=F) + theme_bw() + theme(panel.grid = element_blank(), axis.title = element_blank(),axis.text = element_blank(),axis.ticks = element_blank())

```