---
title: TAP-seq permits efficient expression profiling of target genes in single cells
  (Figure 1d,e)
author: "Schraivogel, Gschwind et al."
date: "02/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
options(bitmapType = "cairo")
```

This vignette reproduces the analysis in Figure 1d,e of Schraivogel, Gschwind, et al. 

## Data

We need downsampled data for the deeply sequenced experiment of the chromosome 11 panel, and the deeply sequenced whole transcriptome experiment. Data were downsampeld at the level of raw reads to realistically simulate a lower sequencing depth. The downsampling can be reproduced from the files provided on GEO by running the rule Figure1 of https://github.com/argschwind/tapseq_manuscript (documented there); alternatively we provide the results - please run the following code in the folder containing this vignette:

```{sh getReferenceData, eval = F}
mkdir -p ../data/tapk562deep/downsampled
wget http://steinmetzlab.embl.de/TAPdata/tapk562deep.zip
unzip -j -d ../data/tapk562deep/downsampled tapk562deep.zip

mkdir -p ../data/wtxk562deep/downsampled
wget http://steinmetzlab.embl.de/TAPdata/wtxk562deep.zip
unzip -j -d ../data/wtxk562deep/downsampled wtxk562deep.zip

rm tapk562deep.zip wtxk562deep.zip
```

In R, we simply load info on the panel:

```{r loadPanelInfo}
#expression levels of the various panels
knowns <- c("HBG1","HBG2","HBE1","HBD","HBB","GATA1","ZFPM2","MYC")
#panels.in <- read.csv("/g/steinmetz/project/singcellTxn/CRISPRdrop/k562_enhancer_cropseq/tapseq_manuscript/meta_data/target_genes_validation.csv", stringsAsFactors = F)
panels.in <- read.csv(url("http://steinmetzlab.embl.de/TAPdata/target_genes_validation.csv"), stringsAsFactors = F)

panels <- list(
  k562.82.10x.low = panels.in$gene[panels.in$panel == "chr11_hs2" | panels.in$gene %in% knowns]
)

```

## Compute environment

The following packages are required:

```{r pressure}

.libPaths("/g/steinmetz/velten/Software/RLibs-seurat3/")
require(ggplot2)
require(plyr)
require(Matrix)
```

## Figure 1d

Simply count the UMIs observed at various read depths:

```{r countUMI, results = "hide"}

folders <- c(
    "k562.82.10x.low" = "../data/tapk562deep/downsampled/"
)


#For targeted seq data
getUMI <- function(n, frPositive = T){
  cat(n,"\n")
  folder <- folders[[n]]
  
  reads <- as.integer(unique(gsub("dge_(.+)_avg_reads_per_cell.txt","\\1",list.files(folder,"dge_.+_avg_reads_per_cell.txt"))))
  out <- lapply(reads, function(nreads ){
    cat(nreads, "\n")
    tap1.con <- gzfile(sprintf("%s/dge_%d_avg_reads_per_cell.txt",folder,nreads))
    DGE <- read.table(tap1.con, header=T, row.names = 1)
    
  if (n != "k562.whole.10x"){
      panel <- panels[[n]]
      DGE <- DGE[panel,]
  }
    
    if (frPositive) data.frame(experiment = n, panel = n, reads = nreads,
               #UMIperCell = apply(DGE,2,sum, na.rm=T),
               frPositive = apply(DGE>0,1,mean), gene = rownames(DGE)) else data.frame(experiment = n, panel = n, reads = nreads,
                                                                                      UMIperCell = apply(DGE,2,sum, na.rm=T), onPanel =T)
    
    })
  do.call(rbind,out)
}


#for whole Tx data, given a reference panel
getUMI2 <- function(ps, frPositive=T, onPanel=T) {
  folder <- "../data/wtxk562deep/downsampled/"
  reads <- as.integer(unique(gsub("dge_(.+)_avg_reads_per_cell.txt","\\1",list.files(folder,"dge_.+_avg_reads_per_cell.txt"))))

  out <- lapply(reads, function(nreads ){
    cat(nreads, "\n")
    tap1.con <- gzfile(sprintf("%s/dge_%d_avg_reads_per_cell.txt",folder,nreads))
    DGE <- read.table(tap1.con, header=T, row.names = 1)

    out2 <- lapply(ps, function(n) {
      panel <- panels[[n]]
      if (onPanel) DGE <- DGE[panel,]
      if (frPositive) data.frame(experiment = "k562.whole.10x", panel = n, reads = nreads,frPositive = apply(DGE>0,1,mean), gene = rownames(DGE)) else data.frame(experiment = "k562.whole.10x", panel = n, reads = nreads,
                                                                                                                                                                  UMIperCell = apply(DGE,2,sum, na.rm=T), onPanel = onPanel)
                 #UMIperCell = apply(DGE,2,sum, na.rm=T))
    })
    do.call(rbind, out2)

    })


  do.call(rbind,out)
}



UMIstat1 <- lapply(names(folders)[1],getUMI, frPositive=F)
UMIstat1 <- do.call(rbind, UMIstat1)
UMIstat1$onPanel <- T

UMIstat2 <- getUMI2(names(panels)[1], frPositive = F)
#UMIstat2 <- getUMI2("WholeTx",T)
UMIstat3 <- getUMI2(names(panels)[1], frPositive = F, onPanel = F)

UMIstat <- rbind(UMIstat1, UMIstat2, UMIstat3)

ggplot(aes(x = reads, y = UMIperCell, color = onPanel, group = paste(experiment, onPanel)), data= UMIstat) + stat_summary(mapping = aes(linetype = experiment), fun.y = mean, geom = "line") + 
  scale_y_log10() + scale_color_manual(values = c("FALSE" = "blue", "TRUE" = "red"),labels = c("FALSE" = "Genome", "TRUE" = "Target panel"),name = "View") + scale_linetype_discrete(labels = c("k562.82.10x.low" = "TAP-seq", "k562.whole.10x" = "Whole Tx"), name = "Method") +
  theme_bw() + xlab("Average reads per cell") + ylab("Average UMI observed") + scale_x_log10() #scale_x_continuous(limits = c(1000,20000), breaks = seq(1000,20000,by=2000), labels = seq(1,20,by=2))

```

## Figure 1e

Simply count the fraction of cells positive at various read depths:



```{r fractionPositive, results = "hide"}

UMIstat1 <- lapply(names(folders)[1],getUMI)
UMIstat1 <- do.call(rbind, UMIstat1)
UMIstat1$onPanel <- T

UMIstat2 <- getUMI2(names(panels)[1])
UMIstat2$onPanel <- T


UMIstat <- rbind(UMIstat1, UMIstat2)
```

For panel e, we additionally need to know the expression level of the various genes, which we take from non-downsampled data

```{r nondownsampled}

total.nods <- read.csv(url("http://steinmetzlab.embl.de/TAPdata/wtxk562deep.counts.csv"), row.names=1)
#this corresponds to ../data/wtxk562deep/dge.txt if the snakemake pipeline was run.

base_expression <- apply(total.nods[panels[[1]],], 1, mean)
bin_expression <- infotheo::discretize(base_expression, nbins = 4)$X
names(bin_expression) <- names(base_expression)
UMIstat$bin <- bin_expression[as.character(UMIstat$gene)]
UMIstat$genome <- UMIstat$panel == "k562.whole.10x"

frPositive <- UMIstat
```

Plot:

```{r paneld, fig.width=5,fig.height=4}
ggplot(aes(x = reads, y = 100*frPositive), data= frPositive) + stat_summary(aes(color =as.factor(bin), linetype = experiment, group= paste(experiment, bin) ), geom="line", fun.y=median) + 
  scale_linetype_discrete(labels = c("k562.82.10x.low" = "TAP-seq", "k562.whole.10x" = "Whole Tx"),name = "Method") + theme_bw() + xlab("Average reads per cell") + ylab("Average UMI observed") + scale_x_log10() +
   ylab("% of cells positive") + scale_color_manual(values = c("1" = "black","2"="blue","3"="red","4"="orange"), name = "Gene expression\nlevel") + scale_x_log10()#scale_x_continuous(limits = c(1000,20000), breaks = seq(1000,20000,by=2000), labels = seq(1,20,by=2))

```