---
title: "Chromatin analyses enhancer screen"
author: "Andreas Gschwind"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  rmd: "chromatin_analyses_screen.Rmd"
output:
  bookdown::html_document2:
    toc: yes
---

# Goal
In depth chromatin analyses of our enhancer mapping results.

```{css echo=FALSE}
/* Define a margin before every image element */
img {
  margin-top: 3em;
}
```

```{r setup, include=FALSE}
# set global chunk options
knitr::opts_chunk$set(echo = FALSE)

# disable all dplyr progress bars
options(dplyr.show_progress = FALSE)

# attach required packages
library(ggpubr)
library(here)
library(kableExtra)
library(rtracklayer)
library(tidyverse)
```

```{r inputData}
# load processed de output (includes confidence level and distance to TSS)
results <- here(snakemake@input$processed_results) %>%
  read.csv(stringsAsFactors = FALSE) %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)),
         significant = if_else(pval_adj_allTests < 0.05, true = "sig", false = "non_sig"))

# encode chip-seq files
chip_files <- here(snakemake@input$encode_chip)
names(chip_files) <- sub("_.*", "", basename(chip_files))
```

# Chromatin assays
Chromatin assays from ENCODE were downloaded and the fold change over control signal is calculated
for every assessed enhancer. Significant enhancers are then compared to non-significant enhancers
with a similar distance to TSS distribution to identify enrichment of chromatin marks at
functionally verified enhancers.

## Chromatin marsk at enhancers
The fold change over control signal of different enhancer and transcription relevant chromatin marks
is computed for the enhancer of every tested enhancer - target gene pair.
```{r loadChIPdata}
# extract cis enhancer - gene pairs, remove significant pairs that increase gene expression and add
# "chr" to chromosome for use with ENCODE data
cis_enh_perts <- results %>%
  filter(enh_type == "cis", abs(dist_to_tss) >= 1000) %>%
  filter(!(significant == "sig" & manual_lfc > 0)) %>%
    mutate(enh_chr = paste0("chr", enh_chr),
         gene_chr = paste0("chr", gene_chr))

# get enhancer coordinates and convert to GRanges object
enh_coords <- cis_enh_perts %>%
  select(enh_chr, enh_start, enh_end, perturbation) %>%
  distinct() %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
  `names<-`(.$perturbation)  # set names to perturbation id

# load encode chip-seq data
chip_data <- lapply(chip_files, FUN = import, which = range(enh_coords), format = "bigWig")
```

```{r chromAtEnhancers}
# function to calculate average assay signal per enhancer
avg_assay_signal <- function(assay, enh_coords) {
  overlaps <- findOverlaps(enh_coords, assay)
  overlaps <- split(overlaps, from(overlaps))
  names(overlaps) <- names(enh_coords[as.numeric(names(overlaps))])
  sapply(overlaps, function(x) {
    sum(mcols(assay[to(x)])$score * width(assay[to(x)])) / sum(width(assay[to(x)]))
  })
}

# calculate average chromatin signal at enhancers
enh_chrom_signal <- chip_data %>%
  sapply(FUN = avg_assay_signal, enh_coords = enh_coords) %>%
  as.data.frame() %>%
  rownames_to_column(var = "perturbation")
  
# transform to data.frame and merge with cis enhancer - gene pairs
cis_enh_perts <- cis_enh_perts %>%
  select(sample, perturbation, enh_chr, gene, pvalue, prop_grna_hits, dist_to_tss, significant) %>%
  left_join(enh_chrom_signal, by = "perturbation")
```

## Normalize for distance to TSS
Enhancers significantly associated with gene expression changes tend to be located in closer
proximity to the target genes TSS than any enhancer to any gene on average. However, every enhancer
has a large spectrum of distances to TSS to all tested genes within the region. To compare
singificant to non-significant enhancers an negative control sample is generated by randomly
sampling. For each significant enhancer - gene pair, three non-significant enhancer - gene pairs are
drawn, whose enhancers have at least one gene with similar distance to TSS as the significant
enhancer - gene pair. To ensure an independent control sample, pairs involing drawn enhancers are
removed from subsequent sampling. This ensures that no enhancer appears more than once in the
control sample. The number of drawn controls per significant pair is set to 3, because with higher
numbers there are simply not enough non-significant enhancer - gene pairs available for sampling
without duplicating enhancers.

```{r drawCtrlsFunction}
# function to randomly draw n non-significant pairs for a significant association. any enhancers
# from drawn non-significant pairs are excluded from subsequent draws
draw_control_pairs <- function(sig, non_sig, n = 3) {
  
  # get distance to tss bins for which a control pair needs to be drawn (repeat n times)
  bins <- select(sig, enh_chr, dTSS_bin)
  bins <- bins[rep(seq_len(nrow(bins)), each = n), ]
  
  # iteratively select a random negative control for each bin and remove any pairs involving that
  # enhancer from pool of non-significant pairs for any subsequenct draws
  random_ctrls <- data.frame()
  for (i in seq_len(nrow(bins))) {
    
    # get pairs on same chromosome and same distance to TSS bin
    nonSig_bin <- filter(non_sig, enh_chr == bins[i, "enh_chr"],
                         dTSS_bin == bins[i, "dTSS_bin"])
    
    # randomly draw a control pair (if possible)
    if (nrow(nonSig_bin)) {
      
      ctrl <- sample_n(nonSig_bin, size = 1)
      random_ctrls <- rbind(random_ctrls, ctrl)
      
      # remove any pairs involving that enhancer from non_sig
      non_sig <- filter(non_sig, perturbation != ctrl$perturbation)
      
    }else{
      warning("Not enough unique enhancers to draw control!\n", call. = FALSE)
    }
  }
  
  return(random_ctrls)
  
}
```

```{r normalizeDistance}
# create distance to TSS bins
dtss_range <- range(cis_enh_perts$dist_to_tss)
bins <- seq(from = dtss_range[1], to = dtss_range[2], by = 1e4)

# bin distance to tss
cis_enh_perts <- cis_enh_perts %>%
  mutate(dTSS_bin = cut(dist_to_tss, bins))

# extract significant enhancer - gene associations (no GATA1, because it's the only on chrX)
sig_pairs <- cis_enh_perts %>%
  filter(significant == "sig", perturbation != "GATA1")

# get other tested enhancer - gene pairs but whose enhancers appear only in non-sigificant tests
nonSig_pairs <- cis_enh_perts %>%
  filter(significant == "non_sig", !perturbation %in% sig_pairs$perturbation,
         perturbation != "GATA1")

# randomly draw 3 control pairs for each significant enhancer - gene association
set.seed(20190617)
nonSig_ctrls <- draw_control_pairs(sig_pairs, non_sig = nonSig_pairs, n = 3)

# combine significant and control pairs
all_pairs <- bind_rows(sig_pairs, nonSig_ctrls)
```

```{r plotDistTSS, fig.cap=cap, fig.height=4, fig.width=9}
# plot distance to TSS for significant pairs and sampled non-significant control pairs
ggplot(all_pairs, aes(x = dist_to_tss, fill = significant)) +
  facet_wrap(~sample) +
  geom_histogram(bins = 30) +
  labs(x = "Distance to TSS", y = "pairs",
       title = "Distance to TSS of significant vs non-significant control pairs") +
  theme_bw()

# figure caption
cap <- "Distance to TSS distribution of significant enhancers and sampled non-significant controls"
```


## Enrichment of chromatin marks at enhancers
The fold change over control signal is compared between enhancers significantly associated with a
target gene and sampled non-significant control enhancers.
```{r plotChromMarks, warning=FALSE, fig.cap=cap, fig.width=6, fig.height=7}
# extract unique enhancers from gene-enhancer pairs and convert to long format
enh_plot <- all_pairs %>%
  select(-c(sample, gene, pvalue, prop_grna_hits, dist_to_tss, dTSS_bin)) %>%
  distinct() %>%
  gather(key = "assay", value = "fc_over_ctrl", -c(perturbation, enh_chr, significant))

# pairwise comparisons using wilcoxon test and format adjusted p-value for plotting
pw_tests <- compare_means(fc_over_ctrl ~ significant, group.by = "assay", data = enh_plot,
                          method = "wilcox.test", p.adjust.method = "holm") %>%
  mutate(p.adj.plot = if_else(p.adj < 0.001, true = "< 0.001",
                              false = paste("=", round(p.adj, digits = 2))))

# plot difference in enhancer relevant chromatin marks between significant and non-significant hits
ggplot(enh_plot, aes(x = significant, y = fc_over_ctrl + 1, fill = significant)) +
  facet_wrap(~assay, ncol = 2) +
  geom_boxplot(notch = TRUE) +
  stat_pvalue_manual(pw_tests, label = "P adj. {p.adj.plot}", y.position = log10(145),
                     inherit.aes = FALSE) +
  labs(y = "Chromatin FC over control") +
  scale_x_discrete(labels = c("Non significant", "Significant")) +
  scale_y_log10(limits = c(NA, 180)) +
  scale_fill_manual(values = c(non_sig = "darkgray", sig = "steelblue")) +
  theme_bw() + 
  theme(axis.title.x = element_blank(), legend.position = "none", text = element_text(size = 15),
        panel.grid = element_blank())

# figure caption
cap <- "Chromatin marks of significant enhancers vs non-significant control enhancers"
```

<br>

## Chromatin marks vs confidence levels
Enhancer - gene pairs are classified as high-confidence, low-confidence and non-significant and
chromatin activity at the involved enhancer is compared between the classes.

### Plot chromatin activity versus confidence level {-}
```{r chromMarksVsConf, warning=FALSE, fig.cap=cap, fig.width=9, fig.height=3.5}
# add label for confidence level based on prop gRNA hits
all_pairs_conf <- all_pairs %>%
  filter(!is.na(prop_grna_hits)) %>%
  mutate(conf = if_else(significant == "sig" & prop_grna_hits >= 0.5,
                        true = "High confidence", false = "Non significant"),
         conf = if_else(significant == "sig" & prop_grna_hits < 0.5,
                        true = "Low confidence", false = conf))

# convert to long format for plot
all_pairs_conf <- all_pairs_conf %>%
  gather(key = "assay", value = "fc_over_ctrl", -c(1:8, dTSS_bin, conf))

# perform kruskal-wallis test (non-parametric one-way anova) per assay and reformat adjusted p-value
kw_tests <- compare_means(formula = fc_over_ctrl ~ conf, method = "kruskal.test",
                          data = all_pairs_conf, group.by = "assay") %>%
  mutate(p.adj.plot = if_else(p.adj < 0.001, true = "< 0.001",
                              false = paste("=", round(p.adj, digits = 2)))) %>%
  mutate(group1 = "High confidence", group2 = "Low confidence")  # add fake groups for text in plots

# plot chromatin signal for high-, low-confidence and non significant enhancer - gene pairs
all_pairs_conf %>%
  filter(assay != "H3K27me3") %>%
  ggplot(., aes(x = conf, y = fc_over_ctrl + 1, fill = conf)) +
  facet_wrap(~assay, ncol = 4) +
  geom_boxplot(notch = TRUE) +
  stat_pvalue_manual(data = filter(kw_tests, assay != "H3K27me3"), y.position = 175, size = 5.5,
                     label = "P adj. {p.adj.plot}", remove.bracket = TRUE, inherit.aes = FALSE) +
  labs(y = "FC over control") +
  scale_x_discrete(labels = c("High conf.", "Low conf.", "Non sig.")) +
  scale_fill_manual(values = c("High confidence" = "goldenrod2",
                               "Low confidence" = "darkslategray3", "Non significant" = "gray")) +
  scale_y_log10(limits = c(NA, 180)) +
  theme_bw() +
  theme(axis.title.x = element_blank(), legend.position = "none", text = element_text(size = 20),
        panel.grid = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1))

# figure caption
cap <- "Chromatin mark signal versus enhancer confidence level"

# save plot for manuscript
ggsave(here("results/plots", "chromatin_sig_vs_conf.pdf"), device = "pdf", width = 9,
       height = 3.5)
```

```{r chromMarksVsConfCheck, warning=FALSE, fig.cap=cap, fig.width=2.8, fig.height=3.5}
# plot H3K27me3 vs confidence levels
all_pairs_conf %>%
  filter(assay == "H3K27me3") %>%
  ggplot(., aes(x = conf, y = fc_over_ctrl + 1, fill = conf)) +
  facet_wrap(~assay) +
  geom_boxplot(notch = TRUE) +
  stat_pvalue_manual(data = filter(kw_tests, assay == "H3K27me3"), y.position = 10, size = 5.5,
                     label = "P adj. {p.adj.plot}", remove.bracket = TRUE, inherit.aes = FALSE) +
  labs(y = "FC over control") +
  scale_x_discrete(labels = c("High conf.", "Low conf.", "Non sig.")) +
  scale_fill_manual(values = c("High confidence" = "goldenrod2",
                               "Low confidence" = "darkslategray3", "Non significant" = "gray")) +
  scale_y_log10(limits = c(NA, 10)) +
  theme_bw() +
  theme(axis.title.x = element_blank(), legend.position = "none", text = element_text(size = 20),
        panel.grid = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1))

# figure caption
cap <- "H3k27me3 signal versus enhancer confidence level"

# save plot for manuscript
ggsave(here("results/plots", "h3k27me3_vs_conf.pdf"), device = "pdf", width = 2.8,
       height = 3.5)
```

<br>

### Pairwise tests {-}
Pairwise tests are performed for all comparison using the non-parametric Wilcoxon signed-rank test.

```{r pairwiseTests}
# perform individual pairwise tests
pw_tests <- compare_means(fc_over_ctrl ~ conf, group.by = "assay", data = all_pairs_conf,
                          method = "wilcox.test", p.adjust.method = "holm")

# reformat for printing
pw_tests <- pw_tests %>%
  select(assay, group1, group2, p.format, p.signif, p.adj) %>%
  rename(p.value = p.format)

# create pretty html table
pw_tests %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), full_width = FALSE,
                position = "left")
```

```{r savePwTests, message=FALSE}
# save pairwise tests as .csv file
write.csv(pw_tests, file = here("results/chromatin_pairwise_tests.csv"), row.names = FALSE)
```
