---
title: "Map enhancers"
author: "Andreas Gschwind"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  rmd: "map_enhancers.Rmd"
output:
  bookdown::html_document2:
    toc: yes
---

# Goal
Differential gene expression was used to map enhancers to their target genes.

```{css echo=FALSE}
/* Define a margin before every image element */
img {
  margin-top: 3em;
}
```

```{r setup, include=FALSE}
# set global chunk options
knitr::opts_chunk$set(echo = FALSE)

# attach required packages
library(ggrepel)
library(gridExtra)
library(here)
library(readxl)
library(rtracklayer)
library(tidyverse)
library(tools)
```

```{r loadData}
# files containing number of cell pert perturbation
ncell_files <- here(snakemake@input$ncells)

# set name for each file
samples <- basename(dirname(dirname(ncell_files)))
names(ncell_files) <- basename(ncell_files) %>%
  sub("ncells_", "", .) %>%
  sub("_.*", "", .) %>%
  paste(samples, ., sep = "_")

# files containing differential expression results
de_output_files <- here(snakemake@input$de_output)

# set name for each file
samples <- basename(dirname(dirname(de_output_files)))
names(de_output_files) <- basename(de_output_files) %>%
  sub("output_", "", .) %>%
  sub("_.*", "", .) %>%
  paste(samples, ., sep = "_")

# read files
ncells <- lapply(ncell_files, FUN = read.csv, stringsAsFactors = FALSE)
de_output <- lapply(de_output_files, FUN = read.csv, stringsAsFactors = FALSE)

# convert to one data.frame
ncells <- ncells %>%
  bind_rows(.id = "id") %>%
  separate(id, into = c("sample", "strategy"), sep = "_") %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

de_output <- de_output %>%
  bind_rows(.id = "id") %>%
  separate(id, into = c("sample", "strategy"), sep = "_") %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

# load processed de output (includes confidence level and distance to TSS)
results <- here(snakemake@input$processed_results) %>%
  read.csv(stringsAsFactors = FALSE) %>% 
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

# load dge data
dge_files <- here(snakemake@input$dge)
names(dge_files) <- basename(dirname(dge_files))
dge <- lapply(dge_files, FUN = read.table, header = TRUE, row.names = "GENE", 
              stringsAsFactors = FALSE)
```

# Number of cell per perturbation
The numbers of cells per perturbation are plotted for per sample and strategy.
```{r ncells}
# minimum number of cells per perturbation used for DE tests
min_cells <- snakemake@config$map_enhancers$min_cells %>%
  unlist() %>%
  data.frame(strategy = names(.), min_cells = .)

# calculate percentage of perturbation with at least min_cells cells
cells_perc <- ncells %>%
  group_by(sample, strategy) %>%
  summarize(cells = sum(cells >= min_cells[strategy, "min_cells"]),
            cells_perc = cells / n() * 100)

# plot number of cells pert perturbation
ggplot(ncells, aes(cells)) +
  facet_grid(strategy~sample, scales = "free") +
  geom_histogram(bins = 45) +
  geom_vline(data = min_cells, aes(xintercept = min_cells), color = "firebrick") +
  labs(title = "Cells per perturbation") +
  theme_bw()
```

# Differentially expressed genes

## P-value distribution
```{r pvalueDistr}
# correct for multiple testing across samples using FDR
de_output <- de_output %>%
  group_by(strategy) %>%
  rename(pval_adj_sample = pval_adj_allTests) %>%
  mutate(pval_adj_allTests = p.adjust(pvalue, method = "fdr"))

# plot p-value distribution
ggplot(de_output, aes(pvalue)) +
  facet_grid(strategy~sample, scales = "free") +
  geom_histogram(bins = 100) +
  labs(title = "P-value distribution") +
  theme_bw()
```

## Singificant hits
The numbers and percentages of significant hits per sample and strategy are plotted.
```{r hitsVsFdr, fig.height=7, fig.width=7}
# count the number and proportion of hits across different FDR values
hits <- de_output %>%
  group_by(sample, strategy) %>%
  summarize("0.05" = sum(pval_adj_allTests < 0.05),
            "0.1" = sum(pval_adj_allTests < 0.1),
            "0.15" = sum(pval_adj_allTests < 0.15),
            "0.2" = sum(pval_adj_allTests < 0.2),
            tests = n()) %>%
  gather(key = "fdr", value = "hits", -sample, -strategy, -tests) %>%
  mutate(fdr = as.numeric(fdr)) %>%
  mutate(prop_hits = hits / tests)

# plot number of hits as function of FDR
p1 <- ggplot(hits, aes(x = fdr, y = hits, color = sample, lty = strategy)) +
  geom_point() +
  geom_line() +
  labs(x = "FDR cutoff", y = "Significant tests", title = "Number of significant tests") +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()

# plot proportion of hits as function of FDR
p2 <- ggplot(hits, aes(x = fdr, y = prop_hits, color = sample, lty = strategy)) +
  geom_point() +
  geom_line() +
  labs(x = "FDR cutoff", y = "Significant tests", title = "Proportion of significant tests") +
  scale_y_continuous(labels = scales::percent) +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()

# print plots
grid.arrange(p1, p2, ncol = 1)
```

## Confidence levels
Significant hits (FDR < 0.05) from the per perturbation approach are used to infer enhancer - target
gene pairs. The per gRNA results are used to assign a confidence level to each enhancer - gene pair
based on how many individual gRNAs had a significant association at FDR < 0.05.
```{r confidence}
# add variable for significant associations
results <- results %>%
  mutate(significant = if_else(pval_adj_allTests < 0.05, true = "sig", false = "non_sig"))

# calculate number of hits per confidence level
hits_conf_levels <- results %>%
  group_by(sample, grna_hits) %>%
  summarize(hits = sum(significant == "sig"))

# plot number of hits per confidence level
ggplot(hits_conf_levels, aes(x = as.factor(grna_hits), y = hits, fill = sample)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "gRNA hits (pvalue adj. < 0.05)", y = "Hits", title = "Hits per confidence level",
       fill = "Sample") +
  scale_fill_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()
```

# Enhancer perturbations
Data on cis enhancer perturbations (no promoter controls, at least 1kb from TSS) are extracted and
genomic properties such as number of enhacners per gene and distance to TSS are assessed.

## Number of cis-enhancers per gene
```{r enhPerGene1}
# extract cis enhancer - gene pairs
cis_enh_perts <- filter(results, enh_type == "cis", abs(dist_to_tss) >= 1000)

# count the number of enhancers for each gene
enh_per_gene <- cis_enh_perts %>%
  group_by(sample, gene) %>%
  summarize(enh = sum(significant == "sig"))

# compute proportion of genes with at least one discovered enhancer
prop_with_enh <- enh_per_gene %>%
  summarize(enh = sum(enh > 0),
            genes = n(),
            prop_enh = enh / genes)

# print table
knitr::kable(prop_with_enh)

# compute number of genes per enhancer count
enh_counts <- enh_per_gene %>%
  filter(enh > 0) %>%
  group_by(sample, enh) %>%
  summarize(genes = n()) %>%
  spread(key = "sample", value = "genes", fill = 0) %>%
  gather(key = "sample", value = "genes", -enh)

# plot number of genes per enhancer count
ggplot(enh_counts, aes(x = factor(enh, levels = seq_len(max(enh))), y = genes, fill = sample)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Enhancers", y = "Genes", title = "Number of enhancers per gene",
       fill = "Sample") +
  scale_fill_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_x_discrete(drop = FALSE) +
  theme_bw()

# save plot for manuscript
ggsave(here("results/plots", "enhancers_per_gene.pdf"), device = "pdf", width = 7, height = 5)
```

```{r enhPerGene2}
# compute number of genes per enhancer count
enh_counts <- enh_per_gene %>%
  filter(enh > 0) %>%
  group_by(enh) %>%
  summarize(genes = n())

# plot number of genes per enhancer count without grouping per chromosome
ggplot(enh_counts, aes(x = factor(enh, levels = seq_len(max(enh))), y = genes)) +
  geom_bar(stat = "identity", position = "dodge", fill = "gray42") +
  labs(x = "Enhancers", y = "Genes", title = "Number of enhancers per gene",
       fill = "Sample") +
  scale_x_discrete(drop = FALSE) +
  theme_bw()
```

## Number of enhancers vs average gene expression
```{r enhsVsExpr}
# function to extract gene expression from dge data
extract_genex <- function(x, vector_pattern) {
  gene_rows <- grep(rownames(x), pattern = vector_pattern, invert = TRUE)
  x[gene_rows, ]
}

# extract gene expression for both experiments
gene_expr <- lapply(dge, FUN = extract_genex, vector_pattern = snakemake@params$vector_pattern)

# calculate average expression per gene
avg_genex <- gene_expr %>%
  lapply(FUN = rowMeans) %>%
  lapply(FUN = function(x) {
    data.frame(gene = names(x), avg_genex = x, stringsAsFactors = FALSE) 
  }) %>%
  bind_rows(.id = "sample") %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

# add average gene expression to enhancers per gene
enh_per_gene <- enh_per_gene %>%
  left_join(avg_genex, by = c("sample", "gene")) %>%
  filter(enh > 0)

# plot number of enhancers vs average gene expression
ggplot(enh_per_gene, aes(x = avg_genex, y = enh, color = sample, label = gene)) +
  facet_wrap(~sample, scales = "free") +
  geom_text() +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()
```

## P-value vs distance to TSS
```{r dTssVsPval}
# calculate -log10 p-value
cis_enh_perts <- mutate(cis_enh_perts, neg_log10_pvalue = -log10(pvalue))

# plot -log10 p-value as a function of distance to TSS
ggplot(cis_enh_perts, aes(x = dist_to_tss / 1e6, y = neg_log10_pvalue, color = significant)) +
  facet_wrap(~sample, ncol = 1, scales = "free") +
  geom_point(data = filter(cis_enh_perts, significant == "non_sig")) +
  geom_point(data = filter(cis_enh_perts, significant == "sig")) +
  labs(x = "Distance to TSS (Mb)", y = expression(-log[10] ~ p - value)) +
  scale_color_manual(values = c(non_sig = "darkgray", sig = "firebrick3")) +
  theme_bw()
```

```{r dTssVsPvalConf}
# set grna hits to NA for non-significant hits
cis_enh_perts_plot <- cis_enh_perts %>%
  mutate(grna_hits = if_else(significant != "sig", true = as.integer(NA), false = grna_hits),
         prop_grna_hits = if_else(significant != "sig", true = as.numeric(NA),
                                  false = prop_grna_hits))

# same plot, but color according to confidence level
ggplot(cis_enh_perts_plot,
             aes(x = dist_to_tss / 1e6, y = neg_log10_pvalue, color = prop_grna_hits)) +
  facet_wrap(~sample, ncol = 1, scales = "free") +
  geom_point(data = filter(cis_enh_perts_plot, significant == "non_sig")) +
  geom_point(data = filter(cis_enh_perts_plot, significant == "sig")) +
  labs(x = "Distance to TSS (Mb)", y = expression(-log[10] ~ p - value)) +
  scale_color_gradient(low = "blue", high = "orangered") +
  theme_bw()
```

```{r assocVsDist, fig.width=7, fig.height=5}
# get hit closest to TSS
closest_hit <- cis_enh_perts_plot %>%
  filter(significant == "sig") %>%
  slice(which.min(abs(dist_to_tss)))

# get hits further away than the closest hit for plotting
cis_enh_perts_plot <- cis_enh_perts_plot %>%
  filter(abs(dist_to_tss) >= abs(closest_hit$dist_to_tss))

# plot decay of association strength over distance within 1Mb of TSS
cis_enh_perts_plot %>%
  ggplot(., aes(x = abs(dist_to_tss) / 1000, y = neg_log10_pvalue, color = prop_grna_hits)) +
    geom_smooth(data = filter(cis_enh_perts_plot, significant == "sig"),
                method = "loess", se = FALSE, color = "black") +
    labs(x = "Distance to TSS (kb)", y = expression(-log[10] ~ p - value),
         color = "Confidence\n(prop. gRNA hits)",
         title = "Association strength vs distance to TSS") +
    geom_point(data = filter(cis_enh_perts_plot, significant == "not_sig")) +
    geom_point(data = filter(cis_enh_perts_plot, significant == "sig")) +
    scale_x_log10() +
    scale_color_gradient(low = "blue", high = "orangered") +
    theme_bw()

# save plot for manuscript
ggsave(here("results/plots", "enh_pval_vs_dist.pdf"), device = "pdf", width = 7, height = 5)
```

## Distance to TSS across confidence levels
```{r dTssVsConf}
# only select significant enhancer - gene pairs that have a associated confidence score
enh_gene_pairs <- filter(cis_enh_perts, significant == "sig", !is.na(prop_grna_hits))

# plot the distance to TSS vs number of gRNA hits
ggplot(enh_gene_pairs, aes(x = as.factor(grna_hits), y = abs(dist_to_tss) / 1000, color = sample)) +
  geom_hline(yintercept = 1000, lty = "dashed", color = "gray42") +
  geom_jitter(width = 0.2) +
  geom_boxplot(color = "black", width = 0.2, fill = NA, outlier.shape = NA) +
  labs(x = "Confidence level (gRNA hits)", y = "Distance to TSS (kb)",
       title = "Distance to TSS vs confidence level", color = "Sample") +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_y_log10() +
  theme_bw()

# save plot for manuscript
ggsave(here("results/plots", "confidence_vs_dist.pdf"), device = "pdf", width = 7, height = 5)
```

## Distance to TSS vs target gene expression
```{r}
# add average gene expression to enhancer - gene pairs
enh_gene_pairs <- left_join(enh_gene_pairs, avg_genex, by = c("sample", "gene"))

# plot distance to TSS vs average expression
ggplot(enh_gene_pairs, aes(x = abs(dist_to_tss) / 1000, y = avg_genex, color = sample)) +
  geom_point() +
  labs(x = "Distance to TSS (kb)", y = "Average gene expression", color = "Sample",
       title = "Gene expression vs distance to TSS") +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()
```

***

# Source
* <a download="map_enhancers.Rmd" href="`r base64enc::dataURI(file = params$rmd,
    mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>
    