---
title: "Map enhancers"
author: "Andreas Gschwind"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  rmd: "map_enhancers.Rmd"
output:
  bookdown::html_document2:
    toc: yes
---

# Goal
Differential gene expression was used to map enhancers to their target genes.

```{css echo=FALSE}
/* Define a margin before every image element */
img {
  margin-top: 3em;
}
```

```{r setup, include=FALSE}
# set global chunk options
knitr::opts_chunk$set(echo = FALSE)

# attach required packages
library(ggrepel)
library(gridExtra)
library(here)
library(readxl)
library(rtracklayer)
library(tidyverse)
library(tools)
```

```{r loadData}
# files containing number of cell pert perturbation
ncell_files <- here(snakemake@input$ncells)

# set name for each file
samples <- basename(dirname(dirname(ncell_files)))
names(ncell_files) <- basename(ncell_files) %>%
  sub("ncells_MAST_", "", .) %>%
  sub("_.*", "", .) %>%
  paste(samples, ., sep = "_")

# files containing differential expression results
de_output_files <- here(snakemake@input$de_output)

# set name for each file
samples <- basename(dirname(dirname(de_output_files)))
names(de_output_files) <- basename(de_output_files) %>%
  sub("output_MAST_", "", .) %>%
  sub("_.*", "", .) %>%
  paste(samples, ., sep = "_")

# read files
ncells <- lapply(ncell_files, FUN = read.csv, stringsAsFactors = FALSE)
de_output <- lapply(de_output_files, FUN = read.csv, stringsAsFactors = FALSE)

# convert to one data.frame
ncells <- ncells %>%
  bind_rows(.id = "id") %>%
  separate(id, into = c("sample", "strategy"), sep = "_") %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

de_output <- de_output %>%
  bind_rows(.id = "id") %>%
  separate(id, into = c("sample", "strategy"), sep = "_") %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

# load processed de output (includes confidence level and distance to TSS)
results <- here(snakemake@input$processed_results) %>%
  read.csv(stringsAsFactors = FALSE) %>% 
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

# load perturbation status
perturb_status_files <- here(snakemake@input$perturb_status)
names(perturb_status_files) <- basename(dirname(perturb_status_files))
perturb_status <- lapply(perturb_status_files, FUN = read.table, header = TRUE,
                         row.names = "TARGET", stringsAsFactors = FALSE)

# load vector targets
vector_target_files <- here(snakemake@input$vector_targets)
names(vector_target_files) <- sub(".+_(chr\\d+)_.+", "\\1", basename(vector_target_files))
vector_targets <- vector_target_files %>%
  lapply(FUN = read.csv, stringsAsFactors = FALSE) %>%
  bind_rows(.id = "sample")
```

# Differential gene expression analysis
Various aspects of the performed differential gene expression analysis are assessed.

## Number of cells per perturbation
The numbers of cells per tested perturbation are plotted for per sample and strategy.

```{r ncells}
# minimum number of cells per perturbation used for DE tests
min_cells <- snakemake@config$map_enhancers$min_cells %>%
  unlist() %>%
  data.frame(strategy = names(.), min_cells = .)

# calculate percentage of perturbation with at least min_cells cells
cells_perc <- ncells %>%
  group_by(sample, strategy) %>%
  summarize(cells = sum(cells >= min_cells[strategy, "min_cells"]),
            cells_perc = cells / n() * 100)

# plot number of cells pert perturbation
ggplot(ncells, aes(cells)) +
  facet_grid(strategy~sample, scales = "free") +
  geom_histogram(bins = 45) +
  geom_vline(data = min_cells, aes(xintercept = min_cells), color = "firebrick") +
  labs(title = "Cells per perturbation") +
  theme_bw()
```

## P-value distribution
```{r pvalueDistr}
# correct for multiple testing across samples using FDR
de_output <- de_output %>%
  group_by(strategy) %>%
  rename(pval_adj_sample = pval_adj_allTests) %>%
  mutate(pval_adj_allTests = p.adjust(pvalue, method = "fdr"))

# plot p-value distribution
ggplot(de_output, aes(pvalue)) +
  facet_grid(strategy~sample, scales = "free") +
  geom_histogram(bins = 100) +
  labs(title = "P-value distribution") +
  theme_bw()
```

## Singificant hits
The numbers and percentages of significant hits per sample and strategy are plotted.

```{r hitsVsFdr, fig.height=7, fig.width=7}
# count the number and proportion of hits across different FDR values
hits <- de_output %>%
  group_by(sample, strategy) %>%
  summarize("0.05" = sum(pval_adj_allTests < 0.05),
            "0.1" = sum(pval_adj_allTests < 0.1),
            "0.15" = sum(pval_adj_allTests < 0.15),
            "0.2" = sum(pval_adj_allTests < 0.2),
            tests = n()) %>%
  gather(key = "fdr", value = "hits", -sample, -strategy, -tests) %>%
  mutate(fdr = as.numeric(fdr)) %>%
  mutate(prop_hits = hits / tests)

# plot number of hits as function of FDR
p1 <- ggplot(hits, aes(x = fdr, y = hits, color = sample, lty = strategy)) +
  geom_point() +
  geom_line() +
  labs(x = "FDR cutoff", y = "Significant tests", title = "Number of significant tests") +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()

# plot proportion of hits as function of FDR
p2 <- ggplot(hits, aes(x = fdr, y = prop_hits, color = sample, lty = strategy)) +
  geom_point() +
  geom_line() +
  labs(x = "FDR cutoff", y = "Significant tests", title = "Proportion of significant tests") +
  scale_y_continuous(labels = scales::percent) +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()

# print plots
grid.arrange(p1, p2, ncol = 1)
```

## Confidence levels
Significant hits (FDR < 0.05) from the per perturbation approach are used to infer enhancer - target
gene pairs. The per gRNA results are used to assign a confidence level to each enhancer - gene pair
based on how many individual gRNAs had a significant association at FDR < 0.05.

```{r confidence}
# add variable for significant associations
results <- results %>%
  mutate(significant = if_else(pval_adj_allTests < 0.05, true = "sig", false = "non_sig"))

# calculate number of hits per confidence level
hits_conf_levels <- results %>%
  group_by(sample, grna_hits) %>%
  summarize(hits = sum(significant == "sig"))

# plot number of hits per confidence level
ggplot(hits_conf_levels, aes(x = as.factor(grna_hits), y = hits, fill = sample)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "gRNA hits (pvalue adj. < 0.05)", y = "Hits", title = "Hits per confidence level",
       fill = "Sample") +
  scale_fill_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  theme_bw()
```

# Cis-enhancers
Data on cis enhancer perturbations (no promoter controls, at least 1kb from TSS) are extracted and 
further investigated. Genomic properties such as number of enhancers per gene and distance to TSS
are assessed to obtain biological insights.

## Enhancer perturbations
The number of enhancers perturbations per cell and cells per perturbation are plotted as an overview
of the screen experiment.

```{r enhPerts}
# known enhancer perturbations
known_enh <- c("GATA1", "HS2", "MYC", "ZFPM2")

# get all transfected enhancer perturbations
enh_pert_targets <- vector_targets %>%
  select(sample, TARGET) %>%
  distinct() %>%
  filter(grepl(TARGET, pattern = "chr\\d+:.*") | TARGET %in% known_enh) %>%
  mutate(sample = paste0(sub("chr", "", sample), "iScreen1"))

# only retain perturbation status on enhancer perturabtions
enh_perts <- perturb_status %>%
  lapply(FUN = function(x) {
    enh_rows <- rownames(x) %in% enh_pert_targets$TARGET
    x[enh_rows, ]
  })
```

**Cells per perturbation**

```{r cellsPerPert, fig.width=5, fig.height=5}
# count number of cells per perturbation
cells_per_pert <- enh_perts %>%
  lapply(FUN = rowSums) %>%
  lapply(FUN = function(x) data.frame(pert = names(x), cells = x, stringsAsFactors = FALSE) ) %>%
  bind_rows(.id = "sample")

# add transfected, but not detected perturbations (in case there are any...)
cells_per_pert <- cells_per_pert %>%
  full_join(enh_pert_targets, by = c("sample" = "sample", "pert" = "TARGET")) %>%
  mutate(cells = replace_na(cells, replace = 0)) %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))
  
# plot number of cell per perturbation
ggplot(cells_per_pert, aes(x = sample, y = cells, fill = sample, color = sample)) +
  geom_violin() +
  labs(y = "Cells") +
  scale_fill_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_y_continuous(breaks = seq(0, 600, 100)) +
  theme_bw() +
  theme(text = element_text(size = 28), legend.position = "none", axis.title.x = element_blank(),
        panel.grid = element_blank())

# save plot for manuscript
ggsave(here("results/plots", "cells_per_pert.pdf"), device = "pdf", width = 5, height = 5)
```

**Perturbations per cell**

```{r pertsPerCell, fig.width=7, fig.height=5}
# count number of perturbations per cell
perts_per_cell <- perturb_status %>%
  lapply(FUN = colSums) %>%
  lapply(FUN = function(x) data.frame(cell = names(x), perts = x, stringsAsFactors = FALSE) ) %>%
  bind_rows(.id = "sample") %>%
  mutate(sample = paste0("Chr", sub("iScreen1", "", sample)))

# compute histogram
perts_per_cell_hist <- perts_per_cell %>%
  group_by(sample, perts) %>%
  summarize(cells = n()) %>%
  mutate(perts = factor(perts, levels = seq(0, max(.$perts))))

# plot perturbations per cell histogram
ggplot(perts_per_cell_hist, aes(x = perts, y = cells / 1000, fill = sample)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Perturbations per cell", y = "Cells (x 1000)",
       fill = "Region:") +
  scale_fill_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_x_discrete(drop = FALSE) +
  theme_bw() +
  theme(text = element_text(size = 22), legend.text = element_text(size = 25),
        legend.position = "top", panel.grid = element_blank())

# save plot for manuscript
ggsave(here("results/plots", "perts_per_cell.pdf"), device = "pdf", width = 7,  height = 5)
```

## Number of cis-enhancers per gene
Cis enhancer gene pairs are extracted by selecting pairs within the same regions and at least 1kb
away from the target genes TSS. Furthermore, any significant hits that lead to an increase in
expression of their target genes are filtered out, as they probably act through alternative
mechanisms. The number of cis-enhancers per gene are counted and plotted.

```{r enhPerGene, fig.width=5, fig.height=5}
# extract cis enhancer - gene pairs and remove significant pairs that increase gene expression
cis_enh_perts <- results %>%
  filter(enh_type == "cis", abs(dist_to_tss) >= 1000) %>%
  filter(!(significant == "sig" & manual_lfc > 0))

# only retain hits within the same target region (discard out of region control enhancers)
cis_enh_perts <- cis_enh_perts %>%
  filter(enh_chr == sub("Chr", "", sample))

# count the number of enhancers for each gene
enh_per_gene <- cis_enh_perts %>%
  group_by(sample, gene) %>%
  summarize(enh = sum(significant == "sig"))

# compute proportion of genes with at least one discovered enhancer
prop_with_enh <- enh_per_gene %>%
  summarize(enh = sum(enh > 0),
            genes = n(),
            prop_enh = enh / genes)

# print table
knitr::kable(prop_with_enh)

# compute number of genes per enhancer count
enh_counts <- enh_per_gene %>%
  filter(enh > 0) %>%
  group_by(sample, enh) %>%
  summarize(genes = n()) %>%
  spread(key = "sample", value = "genes", fill = 0) %>%
  gather(key = "sample", value = "genes", -enh)

# plot number of genes per enhancer count
ggplot(enh_counts, aes(x = factor(enh, levels = seq_len(max(enh))), y = genes, fill = sample)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Enhancers", y = "Genes",
       fill = "Region:") +
  scale_fill_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_x_discrete(drop = FALSE) +
  theme_bw() +
  theme(text = element_text(size = 25), panel.grid = element_blank(),
        legend.position = c(0.8, 0.85))

# save plot for manuscript
ggsave(here("results/plots", "enhancers_per_gene.pdf"), device = "pdf", width = 5, height = 5)
```

## P-value vs distance to TSS
The negative log10 p-value is plotted as function of distance to TSS to show that strong
associations (high -log10 p-value) are located close to their target genes TSS.

```{r dTssVsPval}
# calculate -log10 p-value
cis_enh_perts <- mutate(cis_enh_perts, neg_log10_pvalue = -log10(pvalue))

# plot -log10 p-value as a function of distance to TSS
ggplot(cis_enh_perts, aes(x = dist_to_tss / 1e6, y = neg_log10_pvalue, color = significant)) +
  facet_wrap(~sample, ncol = 1, scales = "free") +
  geom_point(data = filter(cis_enh_perts, significant == "non_sig")) +
  geom_point(data = filter(cis_enh_perts, significant == "sig")) +
  labs(x = "Distance to TSS (Mb)", y = expression(-log[10] ~ p - value)) +
  scale_color_manual(values = c(non_sig = "darkgray", sig = "firebrick3")) +
  theme_bw()
```

```{r dTssVsPvalConf}
# set grna hits to NA for non-significant hits
cis_enh_perts_plot <- cis_enh_perts %>%
  mutate(grna_hits = if_else(significant != "sig", true = as.integer(NA), false = grna_hits),
         prop_grna_hits = if_else(significant != "sig", true = as.numeric(NA),
                                  false = prop_grna_hits))

# same plot, but color according to confidence level
ggplot(cis_enh_perts_plot,
             aes(x = dist_to_tss / 1e6, y = neg_log10_pvalue, color = prop_grna_hits)) +
  facet_wrap(~sample, ncol = 1, scales = "free") +
  geom_point(data = filter(cis_enh_perts_plot, significant == "non_sig")) +
  geom_point(data = filter(cis_enh_perts_plot, significant == "sig")) +
  labs(x = "Distance to TSS (Mb)", y = expression(-log[10] ~ p - value)) +
  scale_color_gradient(low = "blue", high = "orangered", na.value = "gray69") +
  theme_bw()
```

```{r assocVsDist, fig.width=8, fig.height=5}
# get hit closest to TSS
closest_hit <- cis_enh_perts_plot %>%
  filter(significant == "sig") %>%
  slice(which.min(abs(dist_to_tss)))

# get hits further away than the closest hit for plotting
cis_enh_perts_plot <- cis_enh_perts_plot %>%
  filter(abs(dist_to_tss) >= abs(closest_hit$dist_to_tss))

# plot decay of association strength over distance within 1Mb of TSS
p1 <- ggplot(cis_enh_perts_plot, aes(x = abs(dist_to_tss) / 1e6, y = neg_log10_pvalue,
                                     color = prop_grna_hits)) +
    labs(x = "Distance to TSS (Mb)", y = expression(-log[10] ~ p - value),
         color = "Confidence") +
    geom_point(data = filter(cis_enh_perts_plot, significant == "non_sig"), size = 3.5) +
    geom_point(data = filter(cis_enh_perts_plot, significant == "sig"), size = 3.5) +
    scale_color_gradient(low = "blue", high = "orangered", na.value = "gray69") +
    theme_bw() +
    theme(text = element_text(size = 25), panel.grid = element_blank())

# same plot, but with logarithmic distance
p2 <- ggplot(cis_enh_perts_plot, aes(x = abs(dist_to_tss) / 1e6, y = neg_log10_pvalue,
                                     color = prop_grna_hits)) +
    geom_smooth(data = filter(cis_enh_perts_plot, significant == "sig"),
                method = "loess", se = FALSE, color = "black", lwd = 2) +
    labs(x = "Distance to TSS (Mb)", y = expression(-log[10] ~ p - value)) +
    geom_point(data = filter(cis_enh_perts_plot, significant == "non_sig"), size = 2.5) +
    geom_point(data = filter(cis_enh_perts_plot, significant == "sig"), size = 2.5) +
    scale_x_log10(breaks = c(0.01, 0.1, 1, 10), labels = c("0.01", "0.1", "1", "10")) +
    scale_color_gradient(low = "blue", high = "orangered", na.value = "gray69") +
    theme_bw() +
    theme(text = element_text(size = 20), axis.title = element_blank(), legend.position = "none",
          panel.grid = element_blank(), plot.background = element_rect(fill = NA))

# add logarithmic plot as inset of linear plot
p1 + 
  annotation_custom(
    grob = ggplotGrob(p2),
    xmin = 18,
    xmax = 56,
    ymin = 40,
    ymax = 122
  )

# save plot for manuscript
ggsave(here("results/plots", "enh_pval_vs_dist.pdf"), device = "pdf", width = 8, height = 5)
ggsave(here("results/plots", "enh_pval_vs_dist.png"), device = "png", width = 8, height = 5)
```

## Distance to TSS across confidence levels
From the previous plots, it becomes evident, that high-confidence enhancers are located within 1Mb
of the target TSS. This is investigated further by plotting the number of individual gRNA hits vs.
distance to TSS.
```{r dTssVsConf}
# only select significant enhancer - gene pairs
enh_gene_pairs <- filter(cis_enh_perts, significant == "sig")

# plot the distance to TSS vs number of gRNA hits
ggplot(drop_na(enh_gene_pairs),
       aes(x = as.factor(grna_hits), y = abs(dist_to_tss) / 1000,color = sample)) +
  geom_hline(yintercept = 1000, lty = "dashed", color = "gray42") +
  geom_jitter(width = 0.2) +
  geom_boxplot(color = "black", width = 0.2, fill = NA, outlier.shape = NA) +
  labs(x = "Confidence level (gRNA hits)", y = "Distance to TSS (kb)",
       title = "Distance to TSS vs confidence level", color = "Sample") +
  scale_color_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_y_log10() +
  theme_bw() +
  theme(text = element_text(size = 20), panel.grid = element_blank())

# save plot for manuscript
ggsave(here("results/plots", "confidence_vs_dist.pdf"), device = "pdf", width = 7, height = 5)
```

<br>

Enhancer-gene pairs with a proportion of individual gRNA hits >= 0.5 are considered high-confidence
hits. All these hits occur withing 1Mb of the target gene TSS:

```{r confLevelVsdTSS}
# add label for confidence level based on prop gRNA hits
cis_enh_perts_plot <- cis_enh_perts %>%
  filter(!is.na(prop_grna_hits)) %>%
  mutate(conf = if_else(significant == "sig" & prop_grna_hits >= 0.5,
                        true = "High confidence", false = "Non significant"),
         conf = if_else(significant == "sig" & prop_grna_hits < 0.5,
                        true = "Low confidence", false = conf))

# plot distance to TSS of significant enhacner gene - pairs vs non-significant pairs
ggplot(cis_enh_perts_plot,
       aes(x = fct_relevel(conf, "High confidence", "Low confidence"),
           y = abs(dist_to_tss) / 1000, fill = conf)) +
  geom_hline(yintercept = 1000, lty = "dashed") +
  geom_violin() +
  labs(y = "Distance to TSS (kb)") +
  scale_fill_manual(values = c("High confidence" = "goldenrod2", "Low confidence" = "darkslategray3", 
                               "Non significant" = "gray")) +
  scale_y_log10() +
  theme_bw() +
  theme(legend.position = "none", axis.title.x = element_blank(), text = element_text(size = 20),
        panel.grid = element_blank())

# save plot for manuscript
ggsave(here("results/plots", "dist_sig_vs_nonsig.pdf"), device = "pdf", width = 7, height = 5)
```

## Enhancers and gene landscape
The number of TSSs of other genes between the enhancer and its target gene are computed for all
associations.

### Jumped genes
For each enhancer - target gene association the number of jumped genes is calculated and plotted.

```{r geneAnnotations}
# import gencode hg38 annotations
annot_url <- "ftp://ftp.ensembl.org/pub/release-89/gtf/homo_sapiens/Homo_sapiens.GRCh38.89.chr.gtf.gz"
annot <- import(annot_url, format = "gtf")

# extract exon annotations for protein-coding genes
genes <- annot[annot$type == "exon" & annot$gene_biotype == "protein_coding" &
                 annot$transcript_biotype == "protein_coding"]

# only retain annotations on assessed chromosomes (maybe a little faster to overlap)
enh_chrs <- unique(cis_enh_perts$enh_chr)
genes <- genes[seqnames(genes) %in% enh_chrs]
genes <- split(genes, f = genes$gene_name)
```

```{r jumpedGenes}
# calculate enhancer centers
enh_gene_pairs <- enh_gene_pairs %>%
  mutate(enh_center = round((enh_start + enh_end) / 2))

# create (sorted) interval between enhancer center and target gene tss
coords <- enh_gene_pairs %>%
  select(sample, perturbation, gene, gene_tss, enh_center) %>%
  gather(key = "type", value = "coord", -c(sample, perturbation, gene)) %>%
  group_by(sample, perturbation, gene) %>%
  summarize(start = sort(coord)[1],
            end = sort(coord)[2])

# merge with additional data
coords <- enh_gene_pairs %>%
  select(sample, perturbation, gene, chr = enh_chr, dist_to_tss, neg_log10_pvalue,
         prop_grna_hits, grna_hits) %>%
  right_join(coords, by = c("sample", "perturbation", "gene"))
  
# create GRanges object and overlap with protein-coding gene annotations
overlaps <- coords %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
  findOverlaps(query = ., subject = genes)

# convert to list per hit
overlaps_list <- split(overlaps, f = from(overlaps))

# get unique overlapping gene names (genes that are jumoed by the enhancer)
jumped_genes <- lapply(overlaps_list, FUN = function(x) {
  target_gene <- unique(coords[from(x), "gene"])
  all_genes <- names(genes[to(x)])
  all_genes[all_genes != target_gene]
})

# count number of jumped genes per association and add that number to data.frame
n_jumped <- sapply(jumped_genes, FUN = length)
coords$jumped_genes <- n_jumped

# compute histogram
jumped_hist <- coords %>%
  count(jumped_genes) %>%
  mutate(perc = n / sum(n))

# plot the number of jumped genes
total_value = sum(jumped_hist$n)
ggplot(jumped_hist, aes(x = jumped_genes, y = n)) +
  geom_bar(stat = "identity") +
  labs(x = "Genes jumped") +
  scale_y_continuous(sec.axis = sec_axis(~(./total_value), labels = scales::percent,
                                         name = "Percentage"), name = "Counts") +
  theme_bw() +
  theme(text = element_text(size = 15), panel.grid = element_blank())

# save plot for manuscript
ggsave(here("results/plots", "jumped_genes.pdf"), device = "pdf", width = 7, height = 5)

# plot association strength vs jumped genes
ggplot(coords, aes(x = jumped_genes, y = neg_log10_pvalue, color = prop_grna_hits)) +
  geom_point() +
  scale_color_gradient(low = "blue", high = "orangered") +
  labs(x = "Genes jumped", y = expression(-log[10] ~ p - value),
       title = "Association strength vs jumped genes", color = "Confidence\n(prop. gRNA hits)") +
  theme_bw() +
  theme(text = element_text(size = 15), panel.grid = element_blank())

# save plot for manuscript
ggsave(here("results/plots", "pvalue_vs_jumped_genes.pdf"), device = "pdf", width = 7, height = 5)
```

### Nearest gene
Every significant pair is assessed whether the target TSS is the nearest TSS to the enhancer.

```{r nearestTSS, fig.width=5, fig.height=3}
# create GRanges object with enhancer center coordinates for every detected association
enh_coords <- enh_gene_pairs %>%
  select(sample, perturbation, gene, enh_chr, enh_center, dist_to_tss, neg_log10_pvalue,
         grna_hits, prop_grna_hits) %>%
  makeGRangesFromDataFrame(seqnames.field = "enh_chr", start.field = "enh_center",
                           end.field = "enh_center", keep.extra.columns = TRUE)

# function to get a genes TSS
get_tss <- function(x) {
  if (all(strand(x) == "+")) {
    tss <- min(start(x))
  }else if (all(strand(x) == "-")) {
    tss <- max(end(x))
  }else{
    warning("Inconsistent strand information for gene:", unique(x$gene_name), call. = FALSE)
    return(GRanges())
  }
  GRanges(seqnames = unique(seqnames(x)), ranges = IRanges(tss, tss), strand = unique(strand(x)))
}

# get TSS for all genes on chr8 and chr11
gene_tss <- unlist(GRangesList(lapply(genes, FUN = get_tss)))

# find nearest gene TSS to each enhancer
nearest_idx <- nearest(x = enh_coords, subject = gene_tss, ignore.strand = TRUE)
nearest_genes <- names(gene_tss[nearest_idx])

# add to associations
nearest_per_hit <- data.frame(mcols(enh_coords), nearest_gene = nearest_genes,
                              stringsAsFactors = FALSE)

# add variable to specify if closest gene ot the enhancer is the inferred target
nearest_per_hit <- nearest_per_hit %>%
  mutate(is_nearest = if_else(gene == nearest_gene, true = 1, false = 0))

# number and percentage of enhancers where the target gene is the nearest gene
n_nearest <- sum(nearest_per_hit$is_nearest)
perc_is_nearest <- mean(nearest_per_hit$is_nearest) * 100

# labels for x-axis for plot(s)
x_labels <- c(paste0("Nearest TSS (n = ", n_nearest,")"),
              paste0("Other TSS (n = ", nrow(nearest_per_hit) - n_nearest, ")"))

# title for plot
plot_title = paste0("Nearest TSS (", round(perc_is_nearest, digits = 1), "% of cases)")

# plot the number of pairs, where the target TSS is the closest TSS vs. not 
ggplot(nearest_per_hit, aes(x = fct_relevel(factor(is_nearest), "1"), fill = sample)) +
  geom_bar(position = "dodge") +
  labs(title = plot_title, y = "ETPs", fill = "Region") +
  scale_fill_manual(values = c("Chr11" = "indianred3", "Chr8" = "steelblue")) +
  scale_x_discrete(labels = x_labels) +
  theme_bw() +
  theme(axis.title.x = element_blank(), text = element_text(size = 13),
        panel.grid = element_blank())

# save plot for manuscript
ggsave(here("results/plots", "nearest_tss.pdf"), device = "pdf", width = 5, height = 3)
```

**Nearest TSS vs. p-value and confidence level**

```{r nearestTssStats, fig.height=5, fig.width=3.5}
# compare p-value and confidence level between hits where the target gene is the nearest gene and
# hits where this is not the case
nearest_per_hit %>%
  select(is_nearest, neg_log10_pvalue, prop_grna_hits) %>%
  gather(key = "stat", value = "value", -is_nearest) %>%
  ggplot(., aes(x = factor(is_nearest), y = value)) +
  facet_wrap(~stat, ncol = 1, scales = "free") +
  geom_boxplot(fill = "steelblue3") +
  scale_x_discrete(labels = x_labels) +
  theme_bw() +
  theme(axis.title.x = element_blank(), panel.grid = element_blank())
```

***

# Source
* <a download="map_enhancers.Rmd" href="`r base64enc::dataURI(file = params$rmd,
    mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>