---
title: "TAP-seq vs CROP-seq"
author: "Andreas Gschwind"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  rmd: "tapseq_vs_cropseq.Rmd"
output:
  bookdown::html_document2:
    toc: yes
---

# Goal
Compare TAP-seq and CROP-seq.
```{css echo=FALSE}
/* Define a margin before every image element */
img {
  margin-top: 3em;
}
```

```{r setup, include=FALSE}
# set global chunk options
knitr::opts_chunk$set(echo = FALSE)

# attach required packages
library(tidyverse)
library(ggrepel)
library(here)

# create plots directory if needed
dir.create("results/plots", showWarnings = FALSE)
```

# Target gene enrichment
The number of genome reads in the raw data (fastq) was counted for all samples. Similarily the
number of reads aligned to target genes in the final .bam files was counted. The percentage of all
read mapping to target genes is then compared between conventional CROP-seq and TAP-seq.
```{r plotReadsOnTarget, fig.cap=cap, fig.width=5, fig.height=4}
# load read counts
read_counts <- read.csv(here(snakemake@input$reads_on_target), stringsAsFactors = FALSE)

# add variable for approach and rename Sample10X to include assessed chromosome
read_counts_plot <- read_counts %>%
  mutate(approach = if_else(sample == "Sample10X", true = "cropseq", false = "tapseq")) %>%
  mutate(sample = if_else(sample == "Sample10X", true = paste0(sample, " (", panel, ")"),
                          false = sample)) %>%
  arrange(panel, desc(sample))

# plot percentage of total reads on target genes
ggplot(read_counts_plot, aes(x = fct_inorder(sample), y = perc, fill = approach)) +
  geom_bar(stat = "identity") +
  labs(y = "Percentage of reads mapping to target genes", x = NULL) +
  scale_y_continuous(labels = scales::percent) +
  coord_flip() +
  theme_bw() +
  theme(legend.position = "none") +
  scale_fill_manual(values =  c("tapseq" = "firebrick2", "cropseq" = "gray42"))

# figure caption
cap <- "Percentage of reads mapping to assessed target genes."

# save plot to file
ggsave(here("results/plots/reads_on_target.pdf"), device = "pdf", width = 5, height = 4)
```

***

# Number of transcripts per target gene
The average number of detected transcripts per target gene are plotted for TAP-seq and CROP-seq
experiments. CROP-seq experiments were sequenced at much higher depth than TAP-seq experiments,
therefore the downsampled DGE data is used for these plots. This ensures that all experiments have
a comparable sequencing depth of 3500 reads per cell on average.
```{r prepareDGE}
# load tap-seq target genes
target_genes <- read.csv(here(snakemake@input$target_genes), stringsAsFactors = FALSE)

# known enhancer targets for validation
known_enh <- target_genes %>%
  filter(screen == "validation") %>%
  pull(gene)

# get downsampled dge files and set names
ds_dge_files <- here(snakemake@input$dge)
names(ds_dge_files) <- ds_dge_files %>%
  sub("_avg_reads_per_cell.txt", "", .) %>%
  sub(".*data/", "", .) %>%
  sub("/downsampled/dge_", "_", .)

# load downsampled dge data and only retain chr8 and chr11 target genes, convert to long format and
# make one data.frame
ds_dge <- ds_dge_files %>%
  lapply(FUN = read.table, header = TRUE, stringsAsFactors = FALSE) %>%
  lapply(FUN = filter, GENE %in% target_genes$gene) %>%
  lapply(FUN = gather, key = "cell",  value = "txs", -GENE) %>%
  bind_rows(.id = "sample") %>%
  rename(gene = GENE)

# split sample into sample and sequencing depth, and add gene panel and experiment information
ds_dge <- ds_dge %>%
  separate(sample, into = c("sample", "avg_rpc"), sep = "_") %>%
  mutate(avg_rpc = as.numeric(avg_rpc)) %>%
  left_join(target_genes[, c("gene", "panel")], by = "gene") %>%
  mutate(panel = sub("_.+", "", panel)) %>%
  mutate(exp = case_when(sample %in% c("11iv210ng", "11iv22ng") ~ "chr11",
                         sample %in% c("8iv210ng", "8iv22ng") ~ "chr8",
                         TRUE ~ "cropseq"))

# only retain intended target gene expression for tap-seq samples
ds_dge <- ds_dge %>%
  mutate(sample_panel = paste0("chr", sub("iv2.+" ,"", sample))) %>%
  filter(sample_panel == panel | gene %in% known_enh | sample == "Sample10X") %>%
  select(-sample_panel)
```

```{r avgTxs, fig.cap=cap, fig.height=7, fig.width=6}
# only retain data on 3500 reads per cell downsampling
ds_dge_filt <- filter(ds_dge, avg_rpc == 3500)

# calculate average number transcripts per target gene and experiment
avg_txs <- ds_dge_filt %>%
  group_by(sample, exp, gene) %>%
  summarize(avg_txs = mean(txs)) %>%
  mutate(type = if_else(gene %in% known_enh, true = "e-gene", false = "non_e-gene")) %>%
  ungroup()

# extract average transcripts per gene for crop-seq
avg_txs_cropseq <- avg_txs %>%
  ungroup() %>%
  filter(exp == "cropseq") %>%
  select(gene, cropseq = avg_txs)

# add to tap-seq data and calculate log fold change
avg_txs_fc <- avg_txs %>%
  filter(exp != "cropseq") %>%
  left_join(avg_txs_cropseq, by = "gene") %>%
  mutate(fc = avg_txs / cropseq, lfc = log2(fc))

# axis limits
axis_limits <- range(select(avg_txs_fc, cropseq, avg_txs))

# create scatter plots
ggplot(avg_txs_fc, aes(x = cropseq, y = avg_txs, col = type)) + 
  facet_wrap(~sample, ncol = 2, nrow = 2) +
  geom_abline(slope = 1, color = "darkgray", size = 0.5) +
  geom_text_repel(data = filter(avg_txs_fc, type == "e-gene"), aes(label = gene),
                  box.padding = unit(1.5, "lines"), color = "gray50") +
  geom_point() +
  geom_smooth(method = lm,  se = FALSE, color = "steelblue") +
  scale_x_log10(limits = axis_limits) + 
  scale_y_log10(limits = axis_limits) +
  scale_color_manual(values = c("e-gene" = "firebrick2", "non_e-gene" = "gray42")) +
  labs(x = "Average transcripts per gene (CROP-seq)",
       y = "Average transcripts per gene (TAP-seq)") +
  coord_fixed() +
  theme_bw() +
  theme(legend.position = "none")

# figure caption
cap <- "Average transcripts per gene, TAP-seq vs. CROP-seq"

# save plot to file
ggsave(here("results/plots/avg_txs_per_gene.pdf"), device = "pdf", width = 7, height = 7)
```

This comparison is quantified by calculating the log fold change in detected transcripts between the
two methods for every gene:
```{r lfcMethods, fig.cap = cap, fig.height=4, fig.width=5}
# axis limits
lim <- max(abs(avg_txs_fc$lfc))

# plot lfc per gene between methods
ggplot(avg_txs_fc, aes(x = sample, y = lfc, color = type)) +
  geom_hline(yintercept = 0, color = "darkgray", lty = "dashed", size = 1) +
  geom_jitter(width = 0.2, size = 1) +
  geom_boxplot(outlier.shape = NA, fill = NA, color = "gray42", notch = TRUE) +
  labs(x = "TAP-seq sample", y = "Log fold change TAP-seq / CROP-seq") +
  scale_color_manual(values = c("e-gene" = "firebrick2", "non_e-gene" = "gray42")) +
  coord_flip() +
  theme_bw() +
  theme(legend.position = "none")

# figure caption
cap <- "Average LFC TAP-seq vs. CROP-seq"

# save plot to file
ggsave("results/plots/lfc_tapseq_vs_cropseq.pdf", device = "pdf", height = 4, width = 5)
```

***

# Number of detected genes per cell
To assess the sensitivity to detect target genes the percentage of cells in which a gene is detected
is computed. To factor in expression levels of the target genes, the results are binned into
expression quartiles based on CROP-seq expression levels. This analysis is performed for both
TAP-seq and CROP-seq across all downsampled sequencing depths.
```{r detectionSensitivity, fig.cap=cap, fig.height=5, fig.width=7}
# calculate number of cells in which a gene is detected in each experiment
cells_genes <- ds_dge %>%
  group_by(exp, avg_rpc, gene) %>%
  summarize(avg_txs = mean(txs),
            perc_cells = mean(txs > 0))

# compute expression quantiles for all target genes based on average expression in CROP-seq
expr_quantiles <- cells_genes %>%
  ungroup() %>%
  filter(exp == "cropseq", avg_rpc == 20000) %>%
  mutate(quantile = ntile(avg_txs, 4)) %>%
  select(gene, quantile)

# add expression quantile to each gene
cells_genes <- left_join(cells_genes, expr_quantiles, by = "gene")

# reformat for plot
cells_genes_plot <- cells_genes %>%
  ungroup() %>%
  select(-avg_txs) %>%
  spread(key = "exp", value = "perc_cells") %>%
  gather(key = "exp", value = "tapseq", -c(avg_rpc, gene, quantile, cropseq)) %>%
  gather(key = "approach", value = "perc_cells", -c(avg_rpc, gene, quantile, exp)) %>%
  drop_na() %>%
  mutate(avg_rpc = factor(avg_rpc, levels = sort(unique(avg_rpc))),
         approach = fct_relevel(approach, "tapseq"))

# create boxplot comparing detection sensitivity between tap and crop-seq across sequencing depths
ggplot(cells_genes_plot, aes(x = as.factor(quantile), y = perc_cells)) +
  facet_wrap(~exp, ncol = 1, nrow = 2) +
  geom_boxplot(aes(linetype = approach, fill = avg_rpc), outlier.shape = NA) +
  labs(y = "Cells in which genes are detected", x = "CROP-seq expression quantiles",
       fill = "Avg reads per cell", linetype = "Method") +
  scale_y_continuous(labels = scales::percent) +
  theme_bw()

# figure caption
cap <- "Detection sensitivity TAP-seq vs. CROP-seq (all genes should be expressed in K562)"

# save plot to file
ggsave(here("results/plots/detection_sensitivity.pdf"), device = "pdf", height = 5, width = 7)
```

***

# Source
* <a download="tapseq_vs_cropseq.Rmd" href="`r base64enc::dataURI(file = params$rmd,
    mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>
